### PR TITLE
eNikshay: BETS integration case forwarders

### PR DESCRIPTION
**Pull request by NoahCarnahan**
Implements several case forwarders to be used in the eNikshay/BETS integration. There are a couple remaining uncertainties that need to be confirmed in the implementation (marked by TODOs) that should be resolved, hence the "hold off on merge" tag.

Probably best reviewed commit-by-commit.
@proteusvacuum @emord 
buddy @gcapalbo

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 8 errors:

* custom/enikshay/case_utils.py, line 308 - F841 local variable 'e' is assigned to but never used
* custom/enikshay/integrations/bets/models.py, line 2 - F401 'repeater_generators' imported but unused
* custom/enikshay/integrations/bets/repeaters.py, line 9 - F401 'get_prescription_vouchers_from_episode' imported but unused
* custom/enikshay/integrations/bets/repeaters.py, line 60 - E501 line too long (134 > 115 characters)
* custom/enikshay/integrations/bets/tests/test_repeaters.py, line 130 - E231 missing whitespace after ','
* custom/enikshay/integrations/bets/tests/test_repeaters.py, line 231 - E501 line too long (118 > 115 characters)
* custom/enikshay/integrations/bets/views.py, line 162 - E501 line too long (131 > 115 characters)
* custom/enikshay/integrations/bets/views.py, line 186 - E501 line too long (128 > 115 characters)

File: custom/enikshay/tests/utils.py
Comment (NoahCarnahan) [inline custom/enikshay/tests/utils.py]: @proteusvacuum, are prescriptions extensions of episodes?

File: custom/enikshay/tests/utils.py
Comment (NoahCarnahan) [inline custom/enikshay/tests/utils.py]: @proteusvacuum, are vouchers extensions of prescriptions?

File: custom/enikshay/integrations/bets/const.py
Comment (NoahCarnahan) [inline custom/enikshay/integrations/bets/const.py]: @esoergel, are these other event ids documented somewhere?

**Comment by NoahCarnahan** [comment]
whoops, meant to ping @esoergel, not emord.

File: custom/enikshay/integrations/bets/models.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

Comment (NoahCarnahan) [inline custom/enikshay/integrations/bets/models.py]: This is intended

File: custom/enikshay/integrations/bets/repeaters.py
Comment (proteusvacuum) [inline custom/enikshay/integrations/bets/repeaters.py]: don't need this, we migrated.

File: custom/enikshay/integrations/bets/repeater_generators.py
Comment (proteusvacuum) [inline custom/enikshay/integrations/bets/repeater_generators.py]: looks like this `raise` is outside of the `except` block. Is that intended? if so, `msg` will be undefined at this point.

File: custom/enikshay/integrations/bets/repeaters.py
Comment (proteusvacuum) [inline custom/enikshay/integrations/bets/repeaters.py]: in the other repeaters we also make sure this isn't getting sent from a test location (https://github.com/dimagi/commcare-hq/blob/master/custom/enikshay/integrations/utils.py#L21-L30) 

I forgot to include that in the spec... sorry.

File: custom/enikshay/case_utils.py
Comment (proteusvacuum) [inline custom/enikshay/case_utils.py]: we have constants for all the case types in this file

File: custom/enikshay/case_utils.py
Comment (proteusvacuum) [inline custom/enikshay/case_utils.py]: can you add this into the `get_person_case` method above?

File: custom/enikshay/case_utils.py
Comment (proteusvacuum) [inline custom/enikshay/case_utils.py]: I'd also check that each of these is of type `voucher`

File: custom/enikshay/case_utils.py
Comment (proteusvacuum) [inline custom/enikshay/case_utils.py]: Check if these are all of type `prescription`. the `episode` case has a lot of reverse indices pointing to it  of different types.

File: custom/enikshay/const.py
Comment (proteusvacuum) [inline custom/enikshay/const.py]: we have all of these. in an email thread called "(eNikshay - BETS integration) Meeting notes: 30 March".
Email from abhijeet on on April 21

File: custom/enikshay/integrations/bets/repeaters.py
Comment (proteusvacuum) [inline custom/enikshay/integrations/bets/repeaters.py]: Yes plz. (see previous comment)

File: custom/enikshay/integrations/bets/repeaters.py
Comment (proteusvacuum) [inline custom/enikshay/integrations/bets/repeaters.py]: I think you should also check that this is "transitioned" to an outcome with `case_properties_changed`. Otherwise we could get into a situation where we send this multiple times for every form submission against this case until the `event_{}` gets set to sent.

File: custom/enikshay/integrations/bets/repeaters.py
Comment (proteusvacuum) [inline custom/enikshay/integrations/bets/repeaters.py]: also verify this is not a test submission (along with all the other repeaters in this file)

File: custom/enikshay/case_utils.py
Comment (proteusvacuum) [inline custom/enikshay/case_utils.py]: AFAIK `dynamic_case_properties` does a db lookup each time. I think you might need to do this in a loop so you don't do 2 db lookups per loop.

File: custom/enikshay/integrations/bets/models.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

File: custom/enikshay/integrations/bets/repeaters.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/repeaters.py]: F401 'CommCareCase' imported but unused

File: custom/enikshay/integrations/bets/repeaters.py
Comment (proteusvacuum) [inline custom/enikshay/integrations/bets/repeaters.py]: can you make this one of the last checks? Since it potentially does a bunch of db lookups, we would only want it to trigger if all the other cases hold true.

File: custom/enikshay/integrations/bets/repeaters.py
Comment (NoahCarnahan) [inline custom/enikshay/integrations/bets/repeaters.py]: Ah right of course. Removed here: https://github.com/dimagi/commcare-hq/pull/15963/commits/d8fa654d80412dd8f39c125152bfe99b4105b8e4 (also removed from other integrations).

File: custom/enikshay/integrations/bets/repeaters.py
Comment (proteusvacuum) [inline custom/enikshay/integrations/bets/repeaters.py]: check that this case property changed

File: custom/enikshay/tests/utils.py
Comment (proteusvacuum) [inline custom/enikshay/tests/utils.py]: yes, but the identifier is `episode_of_prescription`.

File: custom/enikshay/integrations/bets/models.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

File: custom/enikshay/tests/utils.py
Comment (proteusvacuum) [inline custom/enikshay/tests/utils.py]: yes, but the identifier is `prescription_of_voucher`

**Review by proteusvacuum — CHANGES_REQUESTED** [review]
This is looking awesome!
Couple of comments.
Can you also add the new repeaters to this list:
https://github.com/dimagi/commcare-hq/blob/master/custom/enikshay/reports/repeaters.py#L18-L22
(or somehow change that to use the settings file) so that the eNikshay repeaters get added to our custom report.

File: custom/enikshay/integrations/bets/const.py
Comment (esoergel) [inline custom/enikshay/integrations/bets/const.py]: The `BETS_EVENT_IDS` dict below contains all the event IDs they've sent over.  `VOUCHER_EVENT_ID` is actually two IDs, I think - `chemist_voucher` and `lab_voucher` below.

`AYUSH_REFERRAL_EVENT` corresponds to event 105 below
>  AYUSH/Other provider -Registering and referral of a presumptive TB case in UATBC/e-Nikshay, and patient subsequently gets notified

I don't see an event corresponding to `SUCCESSFUL_TREATMENT_EVENT`, which I believe is this in long form:
> Patients: Cash transfer on successful treatment completion (* trigger - updating of ‘Treatment Outcome’ for an episode )

I'll ping about that one.

Also, the keys in the dict below aren't actually being used for anything - for my part of this I only need a list of valid event IDs.  I stuck the slugs in there as a way to assign names to them.  If you prefer to use individual constants, then we should get rid of that dict and just have a list of the constants so we don't have competing names.

Comment (NoahCarnahan) [inline custom/enikshay/integrations/bets/const.py]: Sounds good thanks Ethan. Should `chemist_voucher`/`lab_voucher` branch on the type of the host case? Or perhaps the `voucher_type` case property? Are the options for this value "prescription" and "test"?

Comment (esoergel) [inline custom/enikshay/integrations/bets/const.py]: `voucher_type` sounds appropriate - the [listed options](https://docs.google.com/spreadsheets/d/1MCG205FOcsYsmKXHoSTjZ6A1iuqCIAESyleBl7G7PR8/edit#gid=0) are "prescription" and "test" which sound like they'd map cleanly to "chemist" and "lab".  I'd want to confirm that with Clayton or Farid first though.

Comment (proteusvacuum) [inline custom/enikshay/integrations/bets/const.py]: correct

File: custom/enikshay/integrations/bets/models.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

File: custom/enikshay/integrations/bets/repeaters.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/repeaters.py]: E501 line too long (117 > 115 characters)

File: custom/enikshay/const.py
Comment (NoahCarnahan) [inline custom/enikshay/const.py]: Ah found that now, thanks

File: custom/enikshay/case_utils.py
Comment (NoahCarnahan) [inline custom/enikshay/case_utils.py]: https://github.com/dimagi/commcare-hq/pull/15963/commits/fbe1cd7019d77d2edfa0ebbb17684978011a3100

File: custom/enikshay/case_utils.py
Comment (NoahCarnahan) [inline custom/enikshay/case_utils.py]: https://github.com/dimagi/commcare-hq/pull/15963/commits/99380802e35ec33b4625d3bbf2b3cb3d29fdbb4e

File: custom/enikshay/case_utils.py
Comment (NoahCarnahan) [inline custom/enikshay/case_utils.py]: https://github.com/dimagi/commcare-hq/pull/15963/commits/99380802e35ec33b4625d3bbf2b3cb3d29fdbb4e

File: custom/enikshay/case_utils.py
Comment (NoahCarnahan) [inline custom/enikshay/case_utils.py]: https://github.com/dimagi/commcare-hq/pull/15963/commits/da865711b29a523193a2365ef2d8d152743fb6d7

File: custom/enikshay/integrations/bets/repeaters.py
Comment (NoahCarnahan) [inline custom/enikshay/integrations/bets/repeaters.py]: https://github.com/dimagi/commcare-hq/pull/15963/commits/a0a3a78e535d35b819d11d5c6f5d3ce56d5f4907

File: custom/enikshay/integrations/bets/repeaters.py
Comment (NoahCarnahan) [inline custom/enikshay/integrations/bets/repeaters.py]: good call, https://github.com/dimagi/commcare-hq/pull/15963/commits/964c0d8e5a702450420ebd59723ddacb43bee6ff

File: custom/enikshay/integrations/bets/repeaters.py
Comment (NoahCarnahan) [inline custom/enikshay/integrations/bets/repeaters.py]: https://github.com/dimagi/commcare-hq/pull/15963/commits/4d69f04879f9095f9bbcbc898f918f3f6c27e8f6

File: custom/enikshay/integrations/bets/repeater_generators.py
Comment (NoahCarnahan) [inline custom/enikshay/integrations/bets/repeater_generators.py]: whoops https://github.com/dimagi/commcare-hq/pull/15963/commits/194b5f225cfc2f546d4396c4709400ee646e5712

File: custom/enikshay/reports/repeaters.py
Comment (proteusvacuum) [inline custom/enikshay/reports/repeaters.py]: yay, thanks.

File: custom/enikshay/case_utils.py
Comment (proteusvacuum) [inline custom/enikshay/case_utils.py]: the others return `case_id`.

File: custom/enikshay/case_utils.py
Comment (proteusvacuum) [inline custom/enikshay/case_utils.py]: would be great to write a test for these accessor methods

**Review by proteusvacuum — COMMENTED** [review]
thanks. few more comments.
bump: https://github.com/dimagi/commcare-hq/pull/15963/files/cf174089dd1416a467bc1182d2719ef219c97179..30a8cf493cb92e480ae1aacb803260f961d948d7#r114140786

File: custom/enikshay/integrations/bets/models.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

File: custom/enikshay/integrations/bets/tests/test_repeaters.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/tests/test_repeaters.py]: F401 'uuid' imported but unused

File: custom/enikshay/integrations/bets/tests/test_repeaters.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/tests/test_repeaters.py]: F401 'create_and_save_a_case' imported but unused

File: custom/enikshay/integrations/bets/models.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

File: custom/enikshay/integrations/utils.py
Comment (NoahCarnahan) [inline custom/enikshay/integrations/utils.py]: @proteusvacuum just a heads up that this affects all the non-BETS repeaters, let me know if this change would impact those negatively.

File: custom/enikshay/integrations/bets/models.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

File: custom/enikshay/integrations/utils.py
Comment (NoahCarnahan) [inline custom/enikshay/integrations/utils.py]: Ha, just saw https://github.com/dimagi/commcare-hq/pull/15998/files, so looks like I do need to change this :)

Comment (proteusvacuum) [inline custom/enikshay/integrations/utils.py]: there's also: https://github.com/dimagi/commcare-hq/blob/master/custom/enikshay/integrations/utils.py#L50-L59

File: custom/enikshay/integrations/bets/models.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

**Comment by NoahCarnahan** [comment]
gah just saw your last comment. Ignore last commit. fixing

File: custom/enikshay/integrations/bets/models.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

File: custom/enikshay/integrations/bets/const.py
Comment (proteusvacuum) [inline custom/enikshay/integrations/bets/const.py]: sp: `EVETN`

File: custom/enikshay/integrations/bets/repeaters.py
Comment (proteusvacuum) [inline custom/enikshay/integrations/bets/repeaters.py]: just to be safe, I think we should also add `case_properties.get(ENROLLED_IN_PRIVATE) == 'true'` here too

File: custom/enikshay/integrations/bets/repeaters.py
Comment (proteusvacuum) [inline custom/enikshay/integrations/bets/repeaters.py]: just to be safe, I think we should also add case_properties.get(ENROLLED_IN_PRIVATE) == 'true' here too

File: custom/enikshay/integrations/bets/repeaters.py
Comment (proteusvacuum) [inline custom/enikshay/integrations/bets/repeaters.py]: this case property is actually called `private_sector_episode_pending_registration`. Sorry about that. I've updated the spec too.

File: custom/enikshay/integrations/bets/repeaters.py
Comment (proteusvacuum) [inline custom/enikshay/integrations/bets/repeaters.py]: check if this is a private sector case

**Review by proteusvacuum — APPROVED** [review]
Yay!
I added a few more comments of things that I didn't add into the spec the first time round (just making sure that when we forward episode and person cases that they were registered in the private sector). Sorry about those.

I think we can probably merge this when those are added and conflicts are fixed, if you feel comfortable with it. 

We can fix the triggers when Akanksha sends us the updated information in a separate PR so this one doesn't get any bigger.

File: custom/enikshay/integrations/bets/models.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

**Comment by NoahCarnahan** [comment]
Thanks for the review @proteusvacuum. Agreed that we should merge this and put the next round of changes in a separate PR. I addressed your latest feedback and build should be done this morning.

**Comment by NoahCarnahan** [comment]
My latest changes broke the tests, but I don't think they are worth fixing since the logic is about to change again anyways. I'll make the next PR be a PR into this one instead.

File: custom/enikshay/integrations/bets/models.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

File: custom/enikshay/integrations/bets/repeater_generators.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/repeater_generators.py]: F401 'get_open_episode_case_from_person' imported but unused
F401 'get_person_case_from_voucher' imported but unused

File: custom/enikshay/integrations/bets/repeaters.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/repeaters.py]: F401 'get_approved_prescription_vouchers_from_episode' imported but unused
F401 'get_episode_case_from_voucher' imported but unused

File: custom/enikshay/integrations/bets/models.py
Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

Comment (dimagimon) [inline custom/enikshay/integrations/bets/models.py]: F401 'repeater_generators' imported but unused

**Comment by NoahCarnahan** [comment]
Do you guys think this should still hold off on merge, or should merge now to avoid future conflicts and such? Sounds like there is still discussion on the spec, but the inconsistencies won't really matter until the repeaters are actually configured for the domain.

**Comment by proteusvacuum** [comment]
Yeah, I think we can merge on ✅

**Pull request merged**
